### ISSUE TITLE
kong.tools.database_cache does not pass on ttl value sent by ldap-auth plugin and custom plugins that use the database_cache module

### ISSUE THREAD

**Issue by cdcurry** [Mashape/kong#1695]
### Summary

The `access` module in the **ldap-auth** plugin caches authenticated credentials using the `kong.tools.database_cache` with
ttl value defined by `conf.cache_ttl`. Full excerpt from the `access` module:

```
local function authenticate(conf, given_credentials)
  local given_username, given_password = retrieve_credentials(given_credentials)
  if given_username == nil then
    return false
  end

  local credential = cache.get_or_set(cache.ldap_credential_key(given_username), function()
    ngx_log(ngx_debug, "[ldap-auth] authenticating user against LDAP server: "..conf.ldap_host..":"..conf.ldap_port)

    local ok, err = ldap_authenticate(given_username, given_password, conf)
    if err ~= nil then ngx_log(ngx_error, err) end
    if not ok then
      return nil
    end
    return {username = given_username, password = given_password}
  end, conf.cache_ttl)

  return credential and credential.password == given_password, credential
end
```

However the ttl value is not use inside `kong.tools.database_cache` when the element is cached. Instead `exptime = 0` is sent to the cache as `ttl` value that in turn is translated into `-1` (=never expires) inside the cache implementation.

The problem is that the credentials never expire from the cache.

Suggested solution is of course to pass on the `ttl`  value to the cache.
### Steps To Reproduce

The issue was discovered while implementing a custom plugin. While looking at how existing plugins in Kong use the `kong.tools.database_cache` with ttl values we found that this was not only an issue with our custom plugin, but a more general issue.
### Additional Details & Logs
- Kong version (`$ kong version`)
  - 0.9.2
- Kong debug-level startup logs (`$ kong start --vv`)
- Kong error logs (`<KONG_PREFIX>/logs/error.log`)
- Kong configuration (registered APIs/Plugins & configuration file)
- Operating System

**Comment by thefosk** [comment]
@shashiranjan84 can you reproduce in a test case?

**Comment by shashiranjan84** [comment]
@thefosk I will check.

**Comment by shashiranjan84** [comment]
Seems like signature of method `get_or_set` has been updated and which broke ttl functionality. I will work on  fix and send a PR soon

**Comment by Tieske** [comment]
Fixed in `next` branch by #1834
